examples/hitl/rearrange_v2/rearrange_v2.py

```diff
@@ -0,0 +1,532 @@
+        assert isinstance(self._gui_agent_ctrl, GuiHumanoidController)
+        base_offset = (
+            self._gui_agent_ctrl.get_articulated_agent().params.base_offset
+        )
+        agent_feet_translation = (
+            self._get_gui_agent_translation() + base_offset
+        )
+        return agent_feet_translation[1]
+
+    def _get_controls_text(self):
+        def get_grasp_release_controls_text():
+            if self._held_obj_id is not None:
+                return "Spacebar: put down\n"
+            elif self._has_grasp_preview:
+                return "Spacebar: pick up\n"
+            else:
+                return ""
+
+        controls_str: str = ""
+        if not self._hide_gui_text:
+            if self._sps_tracker.get_smoothed_rate() is not None:
+                controls_str += f"server SPS: {self._sps_tracker.get_smoothed_rate():.1f}\n"
+            if self._client_display_latency_ms:
+                controls_str += (
+                    f"latency: {self._client_display_latency_ms:.0f}ms\n"
+                )
+            controls_str += "H: show/hide help text\n"
+            controls_str += "P: pause\n"
+            controls_str += "I, K: look up, down\n"
+            controls_str += "A, D: turn\n"
+            controls_str += "W, S: walk\n"
+            if ENABLE_ARTICULATED_OPEN_CLOSE:
+                controls_str += "Z: open/close receptacle\n"
+            controls_str += get_grasp_release_controls_text()
+
+        return controls_str
+
+    def _get_status_text(self):
+        status_str = ""
+
+        if self._paused:
+            status_str += "\n\npaused\n"
+        if self._show_idle_kick_warning:
+            status_str += (
+                "\n\nAre you still there?\nPress any key to keep playing!\n"
+            )
+
+        return status_str
+
+    def _update_help_text(self):
+        controls_str = self._get_controls_text()
+        if len(controls_str) > 0:
+            self._app_service.text_drawer.add_text(
+                controls_str, TextOnScreenAlignment.TOP_LEFT
+            )
+
+        status_str = self._get_status_text()
+        if len(status_str) > 0:
+            self._app_service.text_drawer.add_text(
+                status_str,
+                TextOnScreenAlignment.TOP_CENTER,
+            )
+
+    def _get_agent_pose(self):
+        agent_root = get_agent_art_obj_transform(
+            self.get_sim(), self.get_gui_controlled_agent_index()
+        )
+        return agent_root.translation, agent_root.rotation
+
+    def _get_camera_lookat_pos(self):
+        agent_root = get_agent_art_obj_transform(
+            self.get_sim(), self.get_gui_controlled_agent_index()
+        )
+        lookat_y_offset = mn.Vector3(0, 1, 0)
+        lookat = agent_root.translation + lookat_y_offset
+        return lookat
+
+    def is_user_idle_this_frame(self):
+        return not self._app_service.gui_input.get_any_key_down()
+
+    def _update_for_remote_client_connect_and_idle(self):
```

Comment:
I may be clearer to split this function into connection handling, idle timer handling and sps counter updating.